### ISSUE TITLE
Bug: [return-await] awaited promise allowed in terminal catch block

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#8663]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.4.2&fileType=.tsx&code=IYZwngdgxgBAZgV2gFwJYHsLwBQEoYDeAUDDMgE5iEkwC%2BMUwyUAFtaaeQKbILlbAA7sFTIYABXLoAtqhBcAdNwBWXKMjwBuGrSK0gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6aRfWaJ2gQwDufSvnRgA2pETRoAe2iQANFGa180QrTJ98ZABaQAuuDABfEKaA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
async function f() {
  try {
  } catch {
    return await Promise.reject();
  }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/return-await": ["error", "in-try-catch"]
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

Expected rule to flag on awaited promise in terminal catch block. From the docs: 
> - if you `return` a promise within a `catch`, and there **_is no_** `finally`, then it **_must not_** be `await`ed.

see https://typescript-eslint.io/rules/return-await/#in-try-catch

### Actual Result

no error reported

### Additional Info
 
Strangely, there is actually a _valid_ test case right now that is asserting this incorrect behavior 
https://github.com/typescript-eslint/typescript-eslint/blob/fc08f969e255b55ea35091b297b358bc434a76eb/packages/eslint-plugin/tests/rules/return-await.test.ts#L238-L253
